src/ngx_http_lua_pipe.c

```diff
@@ -0,0 +1,2476 @@
+        }
+
+    } else if (pipe->stderr_ctx->c != NULL) {
+        rev = pipe->stderr_ctx->c->read;
+        return ngx_http_lua_pipe_close_helper(pipe->stderr_ctx, rev, forced);
+    }
+
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_ffi_pipe_proc_shutdown_stdin(ngx_http_lua_ffi_pipe_proc_t *proc,
+    u_char *errbuf, size_t *errbuf_size)
+{
+    ngx_int_t                        rc;
+    ngx_http_lua_pipe_t             *pipe;
+
+    pipe = proc->pipe;
+    if (pipe == NULL) {
+        *errbuf_size = ngx_snprintf(errbuf, *errbuf_size, "closed") - errbuf;
+        return NGX_ERROR;
+    }
+
+    rc = ngx_http_lua_pipe_close_stdin(pipe, 0);
+    if (rc != NGX_OK) {
+        *errbuf_size =
+            ngx_snprintf(errbuf, *errbuf_size, "pipe busy writing") - errbuf;
+        return NGX_ERROR;
+    }
+
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_ffi_pipe_proc_shutdown_stdout(ngx_http_lua_ffi_pipe_proc_t *proc,
+    u_char *errbuf, size_t *errbuf_size)
+{
+    ngx_int_t                        rc;
+    ngx_http_lua_pipe_t             *pipe;
+
+    pipe = proc->pipe;
+    if (pipe == NULL) {
+        *errbuf_size = ngx_snprintf(errbuf, *errbuf_size, "closed") - errbuf;
+        return NGX_ERROR;
+    }
+
+    rc = ngx_http_lua_pipe_close_stdout(pipe, 0);
+    if (rc != NGX_OK) {
+        *errbuf_size =
+            ngx_snprintf(errbuf, *errbuf_size, "pipe busy reading") - errbuf;
+        return NGX_ERROR;
+    }
+
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_ffi_pipe_proc_shutdown_stderr(ngx_http_lua_ffi_pipe_proc_t *proc,
+    u_char *errbuf, size_t *errbuf_size)
+{
+    ngx_int_t                        rc;
+    ngx_http_lua_pipe_t             *pipe;
+
+    pipe = proc->pipe;
+    if (pipe == NULL) {
+        *errbuf_size = ngx_snprintf(errbuf, *errbuf_size, "closed") - errbuf;
+        return NGX_ERROR;
+    }
+
+    if (pipe->merge_stderr) {
+        /* stdout is used internally as stderr when merge_stderr is true */
+        rc = ngx_http_lua_pipe_close_stdout(pipe, 0);
+
+    } else {
+        rc = ngx_http_lua_pipe_close_stderr(pipe, 0);
+    }
+
+    if (rc != NGX_OK) {
+        *errbuf_size =
+            ngx_snprintf(errbuf, *errbuf_size, "pipe busy reading") - errbuf;
+        return NGX_ERROR;
+    }
+
+    return NGX_OK;
+}
+
+
+static void
+ngx_http_lua_pipe_proc_finalize(ngx_http_lua_ffi_pipe_proc_t *proc, int forced)
+{
+    ngx_http_lua_pipe_t          *pipe;
+
+    ngx_log_debug3(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                   "lua pipe finalize process:%p pid:%P forced:%d", proc,
+                   proc->_pid, forced);
+    pipe = proc->pipe;
+
+    if (pipe->node) {
+        ngx_rbtree_delete(&ngx_http_lua_pipe_rbtree, pipe->node);
+        pipe->node = NULL;
+    }
+
+    pipe->dead = 1;
+
+    ngx_http_lua_pipe_close_stdin(pipe, forced);
+    ngx_http_lua_pipe_close_stdout(pipe, forced);
+
+    if (!pipe->merge_stderr) {
+        ngx_http_lua_pipe_close_stderr(pipe, forced);
+    }
+
+    pipe->closed = 1;
```

Comment:
@thibaultcha 
* `dead` is used when the process is gone, but we are still able to read/write to the buffer.
* `closed` is used when the process is 'closed' manually, and we could not read/write to the buffer.

`pipe->dead` could be `1` when `pipe->closed` is `0`.

Reply:
Sorry for the late reply, I just suffered a network problem. The reason why I chose `dead` instead of `exited` is that you could do something with a `dead` process, but you could not do something with an `exited` process. Admittedly, the `pipe->closed` is suitable for the read/write operation, but not for the `wait` operation.